includes/sanitizers/class-amp-link-sanitizer.php

```diff
@@ -0,0 +1,202 @@
+ *  - Forms with internal actions get a hidden 'amp' input added to them.
+ *  - AMP pages get meta[amp-to-amp-navigation] added to them.
+ *  - Any elements in the admin bar are excluded.
+ *
+ * Adapted from https://gist.github.com/westonruter/f9ee9ea717d52471bae092879e3d52b0
+ *
+ * @link https://github.com/ampproject/amphtml/issues/12496
+ * @since 1.4.0
+ */
+class AMP_Link_Sanitizer extends AMP_Base_Sanitizer {
+
+	/**
+	 * Default A2A meta tag content.
+	 *
+	 * @var string
+	 */
+	const DEFAULT_META_CONTENT = 'AMP-Redirect-To; AMP.navigateTo';
+
+	/**
+	 * Placeholder for default args, to be set in child classes.
+	 *
+	 * @var array
+	 */
+	protected $DEFAULT_ARGS = [ // phpcs:ignore WordPress.NamingConventions.ValidVariableName.PropertyNotSnakeCase
+		'paired'       => false, // Only set to true when in a paired mode (will be false when amp_is_canonical()). Controls whether query var is added.
+		'meta_content' => self::DEFAULT_META_CONTENT,
+	];
+
+	/**
+	 * Home host.
+	 *
+	 * @var string
+	 */
+	protected $home_host;
+
+	/**
+	 * Content path.
+	 *
+	 * @var string
+	 */
+	protected $content_path;
+
+	/**
+	 * Admin path.
+	 *
+	 * @var string
+	 */
+	protected $admin_path;
+
+	/**
+	 * Sanitizer constructor.
+	 *
+	 * @param DOMDocument $dom  Document.
+	 * @param array       $args Args.
+	 */
+	public function __construct( DOMDocument $dom, array $args = [] ) {
+		if ( ! isset( $args['meta_content'] ) ) {
+			$args['meta_content'] = self::DEFAULT_META_CONTENT;
+		}
+
+		parent::__construct( $dom, $args );
+
+		$this->home_host    = wp_parse_url( home_url(), PHP_URL_HOST );
+		$this->content_path = wp_parse_url( content_url( '/' ), PHP_URL_PATH );
+		$this->admin_path   = wp_parse_url( admin_url(), PHP_URL_PATH );
+	}
+
+	/**
+	 * Sanitize.
+	 */
+	public function sanitize() {
+		if ( ! empty( $this->args['meta_content'] ) ) {
+			$this->add_a2a_meta( $this->args['meta_content'] );
+		}
+
+		$this->process_links();
+	}
+
+	/**
+	 * Add the amp-to-amp-navigation meta tag.
+	 *
+	 * @param string $content The content for the meta tag, for example 'AMP-Redirect-To; AMP.navigateTo'.
+	 * @return DOMElement|null The added meta element if successful.
+	 */
+	public function add_a2a_meta( $content = self::DEFAULT_META_CONTENT ) {
+		$head = $this->dom->documentElement->getElementsByTagName( 'head' )->item( 0 );
+		if ( ! $head || ! $content ) {
+			return null;
+		}
+		$meta = $this->dom->createElement( 'meta' );
+		$meta->setAttribute( 'name', 'amp-to-amp-navigation' );
+		$meta->setAttribute( 'content', $content );
+		$head->appendChild( $meta );
+		return $meta;
+	}
+
+	/**
+	 * Process links by adding adding AMP query var to links in paired mode and adding rel=amphtml.
+	 */
+	public function process_links() {
+		/**
+		 * Element.
+		 *
+		 * @var DOMElement $element
+		 */
+		$xpath = new DOMXPath( $this->dom );
+
+		// Remove admin bar from DOM to prevent mutating it.
```

Comment:
~Making two mutations here seems wasteful. Wouldn't it be preferable to adapt the XPath queries to skip ancestors of the admin bar instead.~
Some quick googling lets me assume that query would be too expensive.